File: src/lib/panelService.ts
import { ObjectId, WithId } from 'mongodb';
import { SolarPanelConfig } from '@/types';
import { getDb } from './db';

const COLLECTION_NAME = 'paneles';

type SolarPanelDocument = Omit<SolarPanelConfig, '_id' | 'createdAt' | 'updatedAt'> & {
  createdAt: Date;
  updatedAt: Date;
};

function mapPanel(doc: WithId<SolarPanelDocument>): SolarPanelConfig {
  return {
    _id: doc._id.toHexString(),
    manufacturer: doc.manufacturer,
    model: doc.model,
    ratedPowerKw: doc.ratedPowerKw,
    quantity: doc.quantity,
    tiltDegrees: doc.tiltDegrees,
    orientation: doc.orientation,
    createdAt: doc.createdAt.toISOString(),
    updatedAt: doc.updatedAt.toISOString(),
  };
}

function ensurePositive(value: unknown, field: string): number {
  const num = Number(value);
  if (!Number.isFinite(num) || num <= 0) {
    throw new Error(`El campo ${field} debe ser un número mayor que cero.`);
  }
  return num;
}

function ensureNonNegative(value: unknown, field: string): number {
  const num = Number(value);
  if (!Number.isFinite(num) || num < 0) {
    throw new Error(`El campo ${field} debe ser un número mayor o igual que cero.`);
  }
  return num;
}

export async function listPanels(): Promise<SolarPanelConfig[]> {
  const db = await getDb();
  const collection = db.collection<SolarPanelDocument>(COLLECTION_NAME);
  const docs = await collection.find().sort({ updatedAt: -1 }).toArray();
  return docs.map(mapPanel);
}

export async function getPanelById(id: string): Promise<SolarPanelConfig | null> {
  const db = await getDb();
  const collection = db.collection<SolarPanelDocument>(COLLECTION_NAME);
  const _id = new ObjectId(id);
  const doc = await collection.findOne({ _id });
  return doc ? mapPanel(doc) : null;
}

export async function createPanel(payload: Partial<SolarPanelConfig>): Promise<SolarPanelConfig> {
  if (!payload.manufacturer?.trim()) {
    throw new Error('El campo manufacturer es obligatorio.');
  }

  const now = new Date();
  const document: SolarPanelDocument = {
    manufacturer: payload.manufacturer.trim(),
    model: payload.model,
    ratedPowerKw: ensurePositive(payload.ratedPowerKw, 'ratedPowerKw'),
    quantity: Math.trunc(ensurePositive(payload.quantity, 'quantity')),
    tiltDegrees: payload.tiltDegrees !== undefined
      ? ensureNonNegative(payload.tiltDegrees, 'tiltDegrees')
      : undefined,
    orientation: payload.orientation,
    createdAt: now,
    updatedAt: now,
  };

  const db = await getDb();
  const collection = db.collection<SolarPanelDocument>(COLLECTION_NAME);
  const result = await collection.insertOne(document);

  return {
    ...payload,
    _id: result.insertedId.toHexString(),
    manufacturer: document.manufacturer,
    model: document.model,
    ratedPowerKw: document.ratedPowerKw,
    quantity: document.quantity,
    tiltDegrees: document.tiltDegrees,
    orientation: document.orientation,
    createdAt: document.createdAt.toISOString(),
    updatedAt: document.updatedAt.toISOString(),
  };
}

export async function updatePanel(
  id: string,
  payload: Partial<SolarPanelConfig>
): Promise<SolarPanelConfig | null> {
  const db = await getDb();
  const collection = db.collection<SolarPanelDocument>(COLLECTION_NAME);
  const _id = new ObjectId(id);

  const existing = await collection.findOne({ _id });
  if (!existing) {
    return null;
  }

  const update: Partial<SolarPanelDocument> = {};

  if (payload.manufacturer !== undefined) {
    if (!payload.manufacturer.trim()) {
      throw new Error('El campo manufacturer es obligatorio.');
    }
    update.manufacturer = payload.manufacturer.trim();
  }
  if (payload.model !== undefined) update.model = payload.model;
  if (payload.ratedPowerKw !== undefined) {
    update.ratedPowerKw = ensurePositive(payload.ratedPowerKw, 'ratedPowerKw');
  }
  if (payload.quantity !== undefined) {
    update.quantity = Math.trunc(ensurePositive(payload.quantity, 'quantity'));
  }
  if (payload.tiltDegrees !== undefined) {
    update.tiltDegrees = ensureNonNegative(payload.tiltDegrees, 'tiltDegrees');
  }
  if (payload.orientation !== undefined) update.orientation = payload.orientation;

  update.updatedAt = new Date();

  await collection.updateOne({ _id }, { $set: update });
  const refreshed = await collection.findOne({ _id });

  return refreshed ? mapPanel(refreshed) : null;
}

export async function deletePanel(id: string): Promise<boolean> {
  const db = await getDb();
  const collection = db.collection<SolarPanelDocument>(COLLECTION_NAME);
  const _id = new ObjectId(id);
  const result = await collection.deleteOne({ _id });
  return result.deletedCount === 1;
}

File: backend/app/services/panel_service.py
"""
Solar panel CRUD operations stored in MongoDB.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from bson import ObjectId
from pymongo.collection import Collection

from app.database import get_database

COLLECTION_NAME = "paneles"


def _collection() -> Collection:
    return get_database()[COLLECTION_NAME]


def _ensure_positive(value: Any, field: str) -> float:
    if value is None:
        raise ValueError(f"El campo {field} es obligatorio.")
    number = float(value)
    if not (number > 0):
        raise ValueError(f"El campo {field} debe ser un número mayor que cero.")
    return number


def _ensure_non_negative(value: Any, field: str) -> float:
    if value is None:
        raise ValueError(f"El campo {field} es obligatorio.")
    number = float(value)
    if number < 0:
        raise ValueError(f"El campo {field} debe ser un número mayor o igual que cero.")
    return number


def _map_panel(doc: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "_id": str(doc["_id"]),
        "manufacturer": doc.get("manufacturer"),
        "model": doc.get("model"),
        "ratedPowerKw": doc.get("ratedPowerKw"),
        "quantity": doc.get("quantity"),
        "tiltDegrees": doc.get("tiltDegrees"),
        "orientation": doc.get("orientation"),
        "createdAt": doc.get("createdAt").isoformat() if doc.get("createdAt") else None,
        "updatedAt": doc.get("updatedAt").isoformat() if doc.get("updatedAt") else None,
    }


def list_panels() -> List[Dict[str, Any]]:
    cursor = _collection().find().sort("updatedAt", -1)
    return [_map_panel(doc) for doc in cursor]


def _object_id(panel_id: str) -> ObjectId:
    try:
        return ObjectId(panel_id)
    except Exception as exc:  # pragma: no cover - invalid ObjectId
        raise ValueError("Identificador de panel inválido.") from exc


def get_panel(panel_id: str) -> Optional[Dict[str, Any]]:
    doc = _collection().find_one({"_id": _object_id(panel_id)})
    return _map_panel(doc) if doc else None


def _ensure_text(value: Any, field: str) -> str:
    if value is None:
        raise ValueError(f"El campo {field} es obligatorio.")
    text = str(value).strip()
    if not text:
        raise ValueError(f"El campo {field} es obligatorio.")
    return text


def create_panel(payload: Dict[str, Any]) -> Dict[str, Any]:
    manufacturer = _ensure_text(payload.get("manufacturer"), "manufacturer")

    now = datetime.utcnow()
    document: Dict[str, Any] = {
        "manufacturer": manufacturer,
        "model": payload.get("model"),
        "ratedPowerKw": _ensure_positive(payload.get("ratedPowerKw"), "ratedPowerKw"),
        "quantity": int(_ensure_positive(payload.get("quantity"), "quantity")),
        "tiltDegrees": (
            _ensure_non_negative(payload.get("tiltDegrees"), "tiltDegrees")
            if payload.get("tiltDegrees") is not None
            else None
        ),
        "orientation": payload.get("orientation"),
        "createdAt": now,
        "updatedAt": now,
    }

    result = _collection().insert_one(document)
    document["_id"] = result.inserted_id
    return _map_panel(document)


def update_panel(panel_id: str, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    update: Dict[str, Any] = {}

    if "manufacturer" in payload:
        update["manufacturer"] = _ensure_text(payload.get("manufacturer"), "manufacturer")
    if "model" in payload:
        update["model"] = payload.get("model")
    if payload.get("ratedPowerKw") is not None:
        update["ratedPowerKw"] = _ensure_positive(payload["ratedPowerKw"], "ratedPowerKw")
    if payload.get("quantity") is not None:
        update["quantity"] = int(_ensure_positive(payload["quantity"], "quantity"))
    if "tiltDegrees" in payload and payload["tiltDegrees"] is not None:
        update["tiltDegrees"] = _ensure_non_negative(payload["tiltDegrees"], "tiltDegrees")
    if "orientation" in payload:
        update["orientation"] = payload.get("orientation")

    if not update:
        return get_panel(panel_id)

    update["updatedAt"] = datetime.utcnow()
    result = _collection().find_one_and_update(
        {"_id": _object_id(panel_id)},
        {"$set": update},
        return_document=True,
    )
    return _map_panel(result) if result else None


def delete_panel(panel_id: str) -> bool:
    result = _collection().delete_one({"_id": _object_id(panel_id)})
    return result.deleted_count == 1
